packages/rum-core/src/domain/contexts/pageStateContexts.ts

```diff
@@ -0,0 +1,157 @@
+import type { Duration, RelativeTime } from '@datadog/browser-core'
+import {
+  isExperimentalFeatureEnabled,
+  noop,
+  elapsed,
+  addDuration,
+  addEventListeners,
+  addEventListener,
+  relativeNow,
+  DOM_EVENT,
+} from '@datadog/browser-core'
+
+export const PAGE_STATE_TIMELINE_MAX_LENGTH = 2500
+
+export const enum PageState {
+  ACTIVE = 'active',
+  PASSIVE = 'passive',
+  HIDDEN = 'hidden',
+  FROZEN = 'frozen',
+  TERMINATED = 'terminated',
+}
+
+type PageStateTimeline = Array<{ state: PageState; startTime: RelativeTime }>
+
+type Counter = { count: number; duration: Duration }
+export type PageStateContext = {
+  active: Counter
+  passive: Counter
+  hidden: Counter
+  frozen: Counter
+  terminated: Counter
+}
+
+export interface PageStateContexts {
+  getPageStates: (startTime: RelativeTime, duration: Duration) => PageStateContext | undefined
+  stop: () => void
+}
+
+let pageStateTimeline: PageStateTimeline = []
+let state: PageState | undefined
+
+export function startPageStateContexts(): PageStateContexts {
+  if (!isExperimentalFeatureEnabled('resource_page_states')) {
+    return {
+      getPageStates: () => undefined,
+      stop: noop,
+    }
+  }
+
+  addPageState(getState())
+
+  const listenerOpts = { capture: true }
+  const stoppers: Array<{ stop: () => void }> = []
+
+  stoppers.push(
+    addEventListeners(
+      window,
+      [DOM_EVENT.PAGE_SHOW, DOM_EVENT.FOCUS, DOM_EVENT.BLUR, DOM_EVENT.VISIBILITY_CHANGE, DOM_EVENT.RESUME],
+      (ev) => {
+        if (ev.isTrusted) addPageState(getState())
+      },
+      listenerOpts
+    ),
+    addEventListener(
+      window,
+      DOM_EVENT.FREEZE,
+      (ev) => {
+        if (ev.isTrusted) addPageState(PageState.FROZEN)
+      },
+      listenerOpts
+    ),
+    addEventListener(
+      window,
+      DOM_EVENT.PAGE_HIDE,
+      (ev: PageTransitionEvent) => {
+        // If the event's persisted property is `true` the page is about
+        // to enter the back/forward cache, which is also in the frozen state.
+        // If the event's persisted property is not `true` the page is
+        // about to be unloaded.
+        if (ev.isTrusted) addPageState(ev.persisted ? PageState.FROZEN : PageState.TERMINATED)
+      },
+      listenerOpts
+    )
+  )
+
+  function findStartIndex(startTime: RelativeTime) {
+    let i = -1
+    let endTime
+
+    do {
+      i++
+      endTime = i + 1 < pageStateTimeline.length ? pageStateTimeline[i + 1].startTime : Infinity
+    } while (i < pageStateTimeline.length && endTime <= startTime)
+
+    return i
+  }
```

Comment:
<!-- Thoughts represent an idea that popped up from reviewing. These comments are non-blocking by nature, but they are extremely valuable and can lead to more focused initiatives and mentoring opportunities.  -->
**💭 thought:** ‏this is a bit convoluted, we can probably do something simpler